### PR TITLE
Add support for finishing and writing 2D maps.

### PR DESCRIPTION
**Pull request by wohe**
When receiving a FinishTrajectory call, the Cartographer nodes catches up on
computing loop closures, runs a final optimization, and when in 2D writes out
a PGM map similar to map_saver.

### CODE REVIEW

File: cartographer_ros/src/cartographer_node_main.cc [hunk 1]
```diff
@@ -456,6 +466,47 @@ bool Node::HandleSubmapQuery(
   return true;
 }
 
+bool Node::HandleFinishTrajectory(
+    ::cartographer_ros_msgs::FinishTrajectory::Request& request,
+    ::cartographer_ros_msgs::FinishTrajectory::Response& response) {
+  carto::common::MutexLocker lock(&mutex_);
+  map_builder_.sparse_pose_graph()->RunFinalOptimization();
```
Comment (damonkohler) [inline cartographer_ros/src/cartographer_node_main.cc]: You need to call sensor_collator_->FinishTrajectory first.

File: cartographer_ros/src/cartographer_node_main.cc [hunk 2]
```diff
@@ -456,6 +466,49 @@ bool Node::HandleSubmapQuery(
...
+  // TODO(whess): Add multi-trajectory support.
+  sensor_collator_.FinishTrajectory(kTrajectoryBuilderId);
+  map_builder_.sparse_pose_graph()->RunFinalOptimization();
+  // TODO(whess): Write X-rays in 3D.
+  if (options_.map_builder_options.use_trajectory_builder_2d()) {
+    const auto trajectory_nodes =
+        map_builder_.sparse_pose_graph()->GetTrajectoryNodes();
+    if (!trajectory_nodes.empty()) {
+      std::string filename = request.stem + ".pgm";
+      LOG(INFO) << "Saving final map to " << filename << "...";
```
Comment (damonkohler) [inline cartographer_ros/src/cartographer_node_main.cc]: Enclose filename in ''s

File: cartographer_ros/src/cartographer_node_main.cc [hunk 3]
```diff
@@ -456,6 +466,49 @@ bool Node::HandleSubmapQuery(
...
+          if (grid.data[i] >= 0 && grid.data[i] <= 100) {
+            output_file.put((100 - grid.data[i]) * 254 / 100);
+          } else {
+            constexpr uint8_t kUnknownValue = 205;
+            output_file.put(kUnknownValue);
+          }
+        }
+      }
+      output_file.close();
+      CHECK(output_file) << "Writing " << filename << " failed.";
```
Comment (damonkohler) [inline cartographer_ros/src/cartographer_node_main.cc]: Same here.

File: cartographer_ros/src/cartographer_node_main.cc [hunk 4]
```diff
@@ -456,6 +466,49 @@ bool Node::HandleSubmapQuery(
   return true;
 }
 
+bool Node::HandleFinishTrajectory(
+    ::cartographer_ros_msgs::FinishTrajectory::Request& request,
+    ::cartographer_ros_msgs::FinishTrajectory::Response& response) {
+  carto::common::MutexLocker lock(&mutex_);
+  // TODO(whess): Add multi-trajectory support.
+  sensor_collator_.FinishTrajectory(kTrajectoryBuilderId);
```
Comment (damonkohler) [inline cartographer_ros/src/cartographer_node_main.cc]: Have you tested what happens if this is called while sensor data is still coming in? Looking at the code, I would expect any data that arrives after this process begins to be queued up by ROS and then ignored by the collator.

Would be nice to verify that though. A crash would be bad :)

**Comment by wohe** [comment]
PTAL

**Pull request merged**